mcs/tools/mono-api-diff/mono-api-diff.cs

```diff
@@ -50,38 +49,86 @@ static int Main (string [] args)
 				Console.WriteLine ("Available options:");
 				options.WriteOptionDescriptions (Console.Out);
 				Console.WriteLine ();
+				return showHelp ? 0 : 1;
+			}
+
+			TextWriter outputStream = null;
+			try {
+				if (!string.IsNullOrEmpty (output))
+					outputStream = new StreamWriter (output);
+
+				ApiDiff.Generate(extra[0], extra[1], outputStream ?? Console.Out);
+			} catch (Exception e) {
+				Console.WriteLine (e);
 				return 1;
+			} finally {
+				outputStream?.Dispose ();
 			}
+			return 0;
+		}
+	}
+#endif
+
+	public static class ApiDiff
+	{
+		public static void Generate (string firstInfo, string secondInfo, TextWriter outStream)
+		{
+			if (firstInfo == null)
+				throw new ArgumentNullException (nameof (firstInfo));
+			if (secondInfo == null)
+				throw new ArgumentNullException (nameof (secondInfo));
+
+			XMLAssembly ms = CreateXMLAssembly (firstInfo);
+			XMLAssembly mono = CreateXMLAssembly (secondInfo);
 
-			XMLAssembly ms = CreateXMLAssembly (extra [0]);
-			XMLAssembly mono = CreateXMLAssembly (extra [1]);
-			XmlDocument doc = ms.CompareAndGetDocument (mono);
+			Generate (ms, mono, outStream);
+		}
 
-			StreamWriter outputStream = null;
-			if (!string.IsNullOrEmpty (output))
-				outputStream = new StreamWriter (output);
+		public static void Generate (Stream firstInfo, Stream secondInfo, TextWriter outStream)
+		{
+			if (firstInfo == null)
+				throw new ArgumentNullException (nameof (firstInfo));
+			if (secondInfo == null)
+				throw new ArgumentNullException (nameof (secondInfo));
 
-			using (XmlTextWriter writer = new XmlTextWriter (outputStream ?? Console.Out)) {
+			XMLAssembly ms = CreateXMLAssembly (firstInfo);
+			XMLAssembly mono = CreateXMLAssembly (secondInfo);
+
+			Generate (ms, mono, outStream);
+		}
+
+		static void Generate (XMLAssembly first, XMLAssembly second, TextWriter outStream)
+		{
+			if (first == null)
+				throw new ArgumentNullException (nameof (first));
+			if (second == null)
+				throw new ArgumentNullException (nameof (second));
+			if (outStream == null)
+				throw new ArgumentNullException (nameof (outStream));
+
+			XmlDocument doc = first.CompareAndGetDocument (second);
+
+			using (XmlTextWriter writer = new XmlTextWriter (outStream)) {
 				writer.Formatting = Formatting.Indented;
 				doc.WriteTo (writer);
 			}
-
-			return 0;
 		}
 
 		static XMLAssembly CreateXMLAssembly (string file)
+		{
+			using (var stream = File.OpenRead(file)) {
+				return CreateXMLAssembly (stream);
+			}
+		}
+
+		static XMLAssembly CreateXMLAssembly (Stream stream)
 		{
 			XmlDocument doc = new XmlDocument ();
-			doc.Load (File.OpenRead (file));
+			doc.Load (stream);
 
 			XmlNode node = doc.SelectSingleNode ("/assemblies/assembly");
 			XMLAssembly result = new XMLAssembly ();
-			try {
-				result.LoadData (node);
-			} catch (Exception e) {
-				Console.Error.WriteLine ("Error loading {0}: {1}\n{2}", file, e.Message, e);
-				Environment.Exit (1);
```

Comment:
What is the point of catching the exception and then quitting anyway? Just throw and let the caller handle it - chances are they will do a better job.